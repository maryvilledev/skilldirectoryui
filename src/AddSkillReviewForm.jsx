import React, { PropTypes } from 'react';
import axios from 'axios';
import { Button,
  Col,
  ControlLabel,
  Form,
  FormControl,
  FormGroup,
  ButtonToolbar } from 'react-bootstrap';

const validateInput = (userInput) => {
  if (userInput.body === '') {
    alert('Cannot submit empty review.');
    return false;
  }
  if (userInput.team_member_id === '') {
    alert('Please select a Team Member.');
    return false;
  }
  return true;
};

class AddSkillReviewForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      teamMembers: null,
      team_member_id: 0,
      positive: true,
      body: '',
    };
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.onPositiveChange = this.onPositiveChange.bind(this);
    this.onBodyChange = this.onBodyChange.bind(this);
    this.onTeamMemberChange = this.onTeamMemberChange.bind(this);
  }

  componentDidMount() {
    axios.get(`${this.props.api}/teammembers/`)
      .then((result) => {
        this.setState({ teamMembers: result.data });
        console.log(result.data)
      })
      .catch((err) => {
        console.log(`Error GETing team members: ${err}`);
      });
  }

  onChange(key) {
    return (value) => {
      this.setState({
        [key]: value,
      });
    };
  }

  onPositiveChange(ev) {
    // ev.target.value is guaranteed to be a string type so we have to check
    // the contents of ev.target.value
    // I would like to throw an error here since there is something WRONG
    // if ev.target.value is neither 'true' or 'false'
    const isPositive = ev.target.value === 'true';
    this.onChange('positive')(isPositive);
  }

  onBodyChange(ev) {
    this.onChange('body')(ev.target.value);
  }

  onTeamMemberChange(ev) {
    this.onChange('team_member_id')(ev.target.value);
  }

  onSubmit(ev) {
    ev.preventDefault();
    // Error Checking
    if (!validateInput(this.state)) {
      return;
    }
    // Call the onSubmit callback to let Skills handle submitting the review
    this.props.onSubmit(this.state);
  }

  render() {
    let teamMembers = null;
    if (this.state.teamMembers != null) {
      teamMembers = this.state.teamMembers.map((teamMember) => {
        return (
          <option
            key={teamMember.ID}
            value={teamMember.ID}
          >
            {teamMember.name}
          </option>
        );
      });
    }
    return (
      <div>
        <h1>Add Skill Review</h1>
        <Form horizontal onSubmit={this.onSubmit}>
          <FormGroup controlId="teamMember">
            <Col componentClass={ControlLabel} sm={4}>
              Select a Team Member:
            </Col>
            <Col sm={8}>
              <FormControl
                name="teamMemberSelect"
                componentClass="select"
                onChange={this.onTeamMemberChange}
              >
                <option key={0} value={null} />
                {teamMembers}
              </FormControl>
            </Col>
          </FormGroup>
          <FormGroup controlId="reviewPositive">
            <Col componentClass={ControlLabel} sm={4}>
              Is This a Positive Review?
            </Col>
            <Col sm={8}>
              <FormControl
                name="positiveSelect"
                componentClass="select"
                onChange={this.onPositiveChange}
              >
                <option value="true">Yes</option>
                <option value="false">No</option>
              </FormControl>
            </Col>
          </FormGroup>
          <FormGroup controlId="reviewBody">
            <Col componentClass={ControlLabel} sm={2}>
              Review:
            </Col>
            <Col sm={10}>
              <FormControl
                name="reviewBody"
                componentClass="textarea"
                onChange={this.onBodyChange}
              />
            </Col>
          </FormGroup>
          <FormGroup>
            <Col smOffset={2}>
              <ButtonToolbar>
                <Button type="submit" bsStyle="primary">
                  Submit
                </Button>
                <Button onClick={this.props.onCancel} bsStyle="info">
                  Cancel
                </Button>
              </ButtonToolbar>
            </Col>
          </FormGroup>
        </Form>
      </div>
    );
  }
}

AddSkillReviewForm.propTypes = {
  api: PropTypes.string.isRequired,
  onCancel: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
};

export default AddSkillReviewForm;
