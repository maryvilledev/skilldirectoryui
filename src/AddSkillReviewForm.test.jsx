import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme';

import AddSkillReviewForm from './AddSkillReviewForm';

describe('<AddSkillReviewForm />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<AddSkillReviewForm />, div);
  });

  it('matches the stored snapshot', () => {
    const wrapper = shallow(<AddSkillReviewForm />);
    expect(wrapper.get(0)).toMatchSnapshot();
  });

  it('changes state when input to "Team Member" field changes', () => {
    const wrapper = mount(<AddSkillReviewForm />);
    expect(wrapper.state().team_member_id).toBe(0);
    const newID = 1234;
    const event = { target: { value: newID } };
    wrapper.find({ name: 'teamMemberSelect' }).simulate('change', event);
    expect(wrapper.state().team_member_id).toBe(newID);
  });

  it('changes state when input to "Positive Review" field changes', () => {
    const wrapper = mount(<AddSkillReviewForm />);
    expect(wrapper.state().positive).toBe(true);
    const newVal = false;
    const event = { target: { value: newVal } };
    wrapper.find({ name: 'positiveSelect' }).simulate('change', event);
    expect(wrapper.state().positive).toBe(newVal);
  });

  it('should be accept Strings as ev.target.value', () => {
    const wrapper = mount(<AddSkillReviewForm />);
    const positiveSelectElement = wrapper.find('[name="positiveSelect"]');
    const falseEvent = {
      target: {
        value: 'false',
      },
    };
    positiveSelectElement.simulate('change', falseEvent);
    expect(wrapper.state().positive).toEqual(false);
    const trueEvent = {
      target: {
        value: 'true',
      },
    };
    positiveSelectElement.simulate('change', trueEvent);
    expect(wrapper.state().positive).toEqual(true);
  });

  it('changes state when text in "Review" text area changes', () => {
    const wrapper = mount(<AddSkillReviewForm />);
    const newBody = 'Little Bo Peep';
    expect(wrapper.state().body).toBe('');
    const event = { target: { value: newBody } };
    wrapper.find({ name: 'reviewBody' }).simulate('change', event);
    expect(wrapper.state().body).toBe(newBody);
  });
});
