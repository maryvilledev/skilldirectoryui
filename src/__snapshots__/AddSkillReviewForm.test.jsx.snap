exports[`<AddSkillReviewForm /> matches the stored snapshot 1`] = `
<div>
  <h1>
    Add Skill Review
  </h1>
  <Form
    bsClass="form"
    componentClass="form"
    horizontal={true}
    inline={false}
    onSubmit={[Function]}>
    <FormGroup
      bsClass="form-group"
      controlId="teamMember">
      <Col
        bsClass="col"
        componentClass={[Function]}
        sm={4}>
        Select a Team Member:
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        sm={8}>
        <FormControl
          bsClass="form-control"
          componentClass="select"
          name="teamMemberSelect"
          onChange={[Function]}>
          <option
            value={null} />
        </FormControl>
      </Col>
    </FormGroup>
    <FormGroup
      bsClass="form-group"
      controlId="reviewPositive">
      <Col
        bsClass="col"
        componentClass={[Function]}
        sm={4}>
        Is This a Positive Review?
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        sm={8}>
        <FormControl
          bsClass="form-control"
          componentClass="select"
          name="positiveSelect"
          onChange={[Function]}>
          <option
            value="true">
            Yes
          </option>
          <option
            value="false">
            No
          </option>
        </FormControl>
      </Col>
    </FormGroup>
    <FormGroup
      bsClass="form-group"
      controlId="reviewBody">
      <Col
        bsClass="col"
        componentClass={[Function]}
        sm={2}>
        Review:
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        sm={10}>
        <FormControl
          bsClass="form-control"
          componentClass="textarea"
          name="reviewBody"
          onChange={[Function]}
          style={
            Object {
              "height": 150,
            }
          } />
      </Col>
    </FormGroup>
    <FormGroup
      bsClass="form-group">
      <Col
        bsClass="col"
        componentClass="div"
        sm={10}
        smOffset={2}>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="primary"
          disabled={false}
          type="submit">
          Submit
        </Button>
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="info"
          disabled={false}
          onClick={undefined}>
          Cancel
        </Button>
      </Col>
    </FormGroup>
  </Form>
</div>
`;
